### PR TITLE
Commonized native and js IR serialization infrastructure.

### PR DESCRIPTION
**Pull request by alexander-gorshenev**

### CODE REVIEW

**Comment by alexander-gorshenev** [comment]
For reference, the corresponding kotlin PR is here: https://github.com/JetBrains/kotlin/pull/2181

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/serialization/KonanIrlinker.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    : KotlinIrLinker(currentModule, logger, builtIns, symbolTable, forwardModuleDescriptor),
+    DescriptorUniqIdAware by KonanDescriptorUniqIdAware {
+
+    private val forwardDeclarations = mutableSetOf<IrSymbol>()
+
+    override val descriptorReferenceDeserializer = KonanDescriptorReferenceDeserializer(currentModule, resolvedForwardDeclarations)
+
+    override fun reader(moduleDescriptor: ModuleDescriptor, uniqId: UniqId) = moduleDescriptor.konanLibrary!!.irDeclaration(uniqId.index, uniqId.isLocal)
+
+    init {
```
Comment (homuroll) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/serialization/KonanIrlinker.kt]: Looks very similar to JS linker. Think how to generalize.

Comment (alexander-gorshenev) [reply backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/serialization/KonanIrlinker.kt]: I've rewritten these pieces a little. It got better.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/serialization/KonanDeclarationTable.kt [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+class KonanDeclarationTable(builtIns: IrBuiltIns, descriptorTable: DescriptorTable):
+    DeclarationTable(builtIns, descriptorTable, KonanMangler) {
+    override var currentIndex = 0L
+
+    init {
+        loadKnownBuiltins()
+    }
+}
+
+// This is what we pre-populate tables with
```
Comment (olonho) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/serialization/KonanDeclarationTable.kt]: Missing dot at the end.

File: backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/serialization/KonanIrModuleSerializer.kt [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+class KonanIrModuleSerializer(
+    logger: LoggingContext,
+    declarationTable: DeclarationTable,
+    bodiesOnlyForInlines: Boolean = false
+) : IrModuleSerializer(logger, declarationTable, KonanMangler, bodiesOnlyForInlines) {
+
+    override fun backendSpecificExplicitRoot(declaration: IrFunction) =
+        declaration.descriptor.annotations.hasAnnotation(RuntimeNames.exportForCppRuntime) ||
+                declaration.descriptor.annotations.hasAnnotation(RuntimeNames.exportForCompilerAnnotation)
```
Comment (SvyatoslavScherbina) [inline backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/serialization/KonanIrModuleSerializer.kt]: `@ExportForCompiler` declarations must not be explicit roots.

**Pull request merged**